sdv/metadata/single_table.py

```diff
@@ -471,14 +478,173 @@ def validate(self):
 
         # Validate columns
         for column, kwargs in self.columns.items():
-            self._append_error(errors, self._validate_column, column, **kwargs)
+            self._append_error(errors, self._validate_column_args, column, **kwargs)
 
         if errors:
             raise InvalidMetadataError(
                 'The following errors were found in the metadata:\n\n'
                 + '\n'.join([str(e) for e in errors])
             )
 
+    def _validate_metadata_matches_data(self, columns):
+        errors = []
+        metadata_columns = self.columns or {}
+        missing_data_columns = set(columns).difference(metadata_columns)
+        if missing_data_columns:
+            errors.append(
+                f'The columns {sorted(missing_data_columns)} are not present in the metadata.')
+
+        missing_metadata_columns = set(metadata_columns).difference(columns)
+        if missing_metadata_columns:
+            errors.append(
+                f'The metadata columns {sorted(missing_metadata_columns)} '
+                'are not present in the data.'
+            )
+
+        if errors:
+            raise InvalidDataError(errors)
+
+    def _get_primary_and_alternate_keys(self):
+        """Get set of primary and alternate keys.
+
+        Returns:
+            set:
+                Set of keys.
+        """
+        keys = set(self.alternate_keys)
+        if self.primary_key:
+            keys.update({self.primary_key})
+
+        return keys
+
+    def _get_set_of_sequence_keys(self):
+        """Get set with a sequence key.
+
+        Returns:
+            set:
+                Set of keys.
+        """
+        if isinstance(self.sequence_key, tuple):
+            return set(self.sequence_key)
+
+        if isinstance(self.sequence_key, str):
+            return {self.sequence_key}
+
+        return set()
+
+    def _validate_keys_dont_have_missing_values(self, data):
+        errors = []
+        keys = self._get_primary_and_alternate_keys()
+        keys.update(self._get_set_of_sequence_keys())
+        for key in sorted(keys):
+            if pd.isna(data[key]).any():
+                errors.append(f"Key column '{key}' contains missing values.")
+
+        return errors
+
+    @staticmethod
+    def _format_invalid_values_string(invalid_values):
+        invalid_values = sorted(invalid_values, key=lambda x: str(x))
+        if len(invalid_values) > 3:
+            return invalid_values[:3] + [f'+ {len(invalid_values) - 3} more']
```

Comment:
A similar thing happens here:
https://github.com/sdv-dev/SDV/blob/6480f1f066ef302d718d147887adda3b588a495f/sdv/constraints/base.py#L204-L210

Maybe we can generalize this function into something that takes a dataframe and number of rows, and returns the string with the top x rows + y more as a string. Then we can move it into utils